File: practices/practice4.1/test/src/PizzaDirectorTest.cpp
#include "MyTest.hpp"
#include "PizzaDirector.hpp"
#include "ConcretePizzaBuilder.hpp"

void PizzaDirectorTest(UnitTests& testSuite) {
    testSuite.addTest("PizzaDirectorMakePizzaTest",
    [](){
        // Build:
        std::map<std::string, std::pair<double, bool>> ingredients = {
              {"Cheese", {1.0, true}},
              {"Pepperoni", {1.5, false}},
              {"Olives", {0.75, false}}
        };

        // Operate:
        std::unique_ptr<PizzaBuilder> builder = std::make_unique<ConcretePizzaBuilder>(ingredients);
        PizzaDirector director(builder);
        builder->addIngredient("Pepperoni", 2);
        std::unique_ptr<Pizza> pizza = director.makePizza();

        // Check:
        ASSERT_EQ(pizza->getIngredientCount("Cheese"), 1)
        ASSERT_EQ(pizza->getIngredientCount("Pepperoni"), 2)
    });
}

File: homeworks/homework4.1/src/test/java/com/pizza/PizzaDirectorTest.java
package com.pizza;

import static org.junit.Assert.*;
import org.junit.Test;
import java.util.Map;

public class PizzaDirectorTest {

    @Test
    public void testMakePizza() {
        Map<String, Ingredient> ingredients = Map.of(
                "Cheese", new Ingredient(1.0, true, 1),
                "Pepperoni", new Ingredient(1.5, false, 1),
                "Olives", new Ingredient(0.75, false, 1)
        );

        PizzaBuilder builder = new ConcretePizzaBuilder(ingredients);
        PizzaDirector director = new PizzaDirector(builder);
        builder.addIngredient("Pepperoni", 2);

        Pizza pizza = director.makePizza();
        assertEquals(1, pizza.getIngredientCount("Cheese"));
        assertEquals(2, pizza.getIngredientCount("Pepperoni"));
    }
}